src/serde_support.rs

```diff
@@ -0,0 +1,259 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+extern crate alloc;
+
+use alloc::format;
+use alloc::string::String;
+
+use dusk_bytes::Serializable;
+use serde::de::Error as SerdeError;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::{
+    PublicKey, PublicKeyDouble, PublicKeyVarGen, SecretKey, SecretKeyVarGen,
+    Signature, SignatureDouble, SignatureVarGen,
+};
+
+impl Serialize for PublicKey {
+    fn serialize<S: Serializer>(
+        &self,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        let s = bs58::encode(self.to_bytes()).into_string();
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> Deserialize<'de> for PublicKey {
+    fn deserialize<D: Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<Self, D::Error> {
+        let s = String::deserialize(deserializer)?;
+        let mut bytes = [0; Self::SIZE];
+        let bytes_written = bs58::decode(&s)
+            .into(&mut bytes)
+            .map_err(SerdeError::custom)?;
+        if bytes_written == Self::SIZE {
+            PublicKey::from_bytes(&bytes)
+                .map_err(|err| SerdeError::custom(format!("{:?}", err)))
+        } else {
+            Err(SerdeError::custom(
+                "Failed to deserialize PublicKey: invalid byte length",
+            ))
+        }
```

Comment:
I'm not really satisfied with the error-handling when deserializing as it will be unclear at which step the deserializing fails if it does.

How about doing it in multiple steps:
```suggestion
        let decoded = bs58::decode(&s).into_vec().map_err(|e| {
            SerdeError::custom("Failed to decode PublicKey: {e}")
        })?;

        let bytes: [u8; Self::SIZE] = decoded
            .try_into()
            .map_err(SerdeError::invalid_length(decoded.len(), Self::SIZE))?;

        Self::from_bytes(&bytes).map_err(|e| {
            SerdeError::custom("Failed to deserialize PublicKey: {e}")
        })
```

Reply:
Yes, that is true. I saw two problems with the approach:
- We use the type name in the error string, that is written without using our own defined error types, which makes it harder to maintain. If this is re-imported as a different type name, or if this is renamed in the future, the error string has to be changed here as well, which can be forgotten.
- I think the error message is already descriptive enough, as a bs58 error on a bs58 input is pretty clear what the problem is/was, with possible additional context in a consuming application on where exactly this error happened.

@moCello let me know if you think otherwise

Reply:
I just checked the serde errors and we could also use invalid_value for the base58 error case. 
```rust
    /// Raised when a `Deserialize` receives a value of the right type but that
    /// is wrong for some other reason.
    ///
    /// The `unexp` argument provides information about what value was received.
    /// This is the value that was present in the input file or other source
    /// data of the Deserializer.
    ///
    /// The `exp` argument provides information about what value was being
    /// expected. This is the type that is written in the program.
    ///
    /// For example if we try to deserialize a String out of some binary data
    /// that is not valid UTF-8, the unexpected value is the bytes and the
    /// expected value is a string.
    #[cold]
    fn invalid_value(unexp: Unexpected, exp: &Expected) -> Self {
        Error::custom(format_args!("invalid value: {}, expected {}", unexp, exp))
    }
```

Reply:
@Neotamandua @moCello, so to confirm:
- the error messages shouldn't have the type names,
- the error "Failed to deserialize PublicKey: {e}" should be more along the lines of "Failed to deserialize: {e}"
- bs58, hex, etc, should have errors reported with `Error::invalid_value`.